substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/NativeImageInlineDuringParsingPlugin.java

```diff
@@ -0,0 +1,548 @@
+    private final HostedProviders providers;
+
+    /* the map in which we place the result of a inlining decision */
+    private static final ConcurrentHashMap<AnalysisMethod, ConcurrentHashMap<Integer, InvocationResult>> dataInline = new ConcurrentHashMap<>();
+
+    public NativeImageInlineDuringParsingPlugin(boolean analysis, HostedProviders providers) {
+        this.analysis = analysis;
+        this.providers = providers;
+    }
+
+    /**
+     * Determines whether a call to a given method is to be inlined.
+     *
+     * @param b the context
+     * @param method the target method of an invoke
+     * @param args the arguments to the invoke
+     */
+    @Override
+    public InlineInfo shouldInlineInvoke(GraphBuilderContext b, ResolvedJavaMethod method, ValueNode[] args) {
+        if (b.parsingIntrinsic()) {
+            /* We are not interfering with any intrinsic method handling. */
+            return null;
+        }
+
+        if (method.getAnnotation(NeverInline.class) != null || method.getAnnotation(NeverInlineTrivial.class) != null) {
+            return null;
+        }
+
+        if (method.getAnnotation(RestrictHeapAccess.class) != null || method.getAnnotation(Uninterruptible.class) != null ||
+                b.getMethod().getAnnotation(RestrictHeapAccess.class) != null || b.getMethod().getAnnotation(Uninterruptible.class) != null) {
+            /*
+             * Caller or callee have an annotation that might prevent inlining. We don't check the
+             * exact condition but instead always bail out for simplicity.
+             */
+            return null;
+        }
+
+        CallSite callSite = new CallSite(toAnalysisMethod(b.getMethod()), b.bci());
+
+        InvocationResult inline;
+        if (analysis) {
+            BigBang bb = ((AnalysisBytecodeParser) b).bb;
+            TrivialMethodDetector detector = new TrivialMethodDetector(bb, providers, ((SharedBytecodeParser) b).getGraphBuilderConfig(), b.getOptions(), b.getDebug());
+            InvocationResult newResult = detector.analyzeMethod(callSite, (AnalysisMethod) method, args, null);
+            putIfAbsent(method, b.bci(), newResult);
+            inline = newResult;
+
+        } else {
+            inline = getResult(method, b.bci());
+        }
+
+        if (inline instanceof InvocationResultInline) {
+            InvocationResultInline inlineData = (InvocationResultInline) inline;
+            VMError.guarantee(inlineData.callee.equals(toAnalysisMethod(method)));
+
+            VMError.guarantee(((SharedBytecodeParser) b).inlineDuringParsingState == null);
+            ((SharedBytecodeParser) b).inlineDuringParsingState = inlineData;
+
+            if (analysis) {
+                AnalysisMethod aMethod = (AnalysisMethod) method;
+                aMethod.registerAsImplementationInvoked(null);
+                if (!aMethod.isStatic()) {
+                    ensureParsed(((AnalysisBytecodeParser) b).bb, aMethod);
```

Comment:
Why do we need this? Again if the method is inlined, do we have to parse it?

Reply:
Yes, I overlooked that. We don't need this. This part of code and method are removed.